OFEP-single-context-paradigm.md

```diff
@@ -0,0 +1,84 @@
+## OFEP-single-context-paradigm
+
+## State: DRAFTING
+
+This draft outlines a "single-context" paradigm - an alternative pattern and set of APIs supporting client use-cases.
+
+## Background
+
+In contrast with server-side or other service-type applications, client side apps typically operate in the context of a single user.
+Most feature flagging SDKs for these applications have been designed with this in mind.
+In summary, most client/web SDKs operate something like this:
+
+- an initialization occurs, which fetches evaluated flags in bulk for a given context (user)
+- the evaluated flags are cached in the SDK
+- flag evaluations take place against this cache, without a need to provide context (context was already used to evaluate flags in bulk)
+- Functions/methods are exposed on the SDK that signal the cache is no longer valid, and must be reconciled based on a context change. This frequently involves a network request or I/O operation.
+
+This paradigm doesn't fit well with the existing SDK, which, like most server-side SDKs, emphasizes realtime evaluation and context-per-evaluation.
+
+Though not all client side SDKs function in this way, those that do allow context per evaluation can conform to this model fairly easily.
+
+## Design
+
+To better support the "single-context" paradigm we need to define some additional handlers and add some flexibility to current APIs.
+In short, the application author now sets context globally using the existing global context mutator on the OpenFeature API object.
+This single context is used by providers for initialization and fetching evaluated flag values in bulk.
+All evaluations functions are "context-less".
+When the context is changed, providers are signalled to update their cache of evaluated flags.
+
+### Provider changes
+
+#### On-context-set handler
+
+Providers must have a mechanism to understand when their cache of evaluated flags must be invalidated or updated. An `on-context-set` handler can be defined which performs whatever operations are needed to reconcile the evaluated flags with the new context. The OpenFeature SDK calls this handler when the global context is modified.
```

Comment:
```suggestion
Providers may need a mechanism to understand when their cache of evaluated flags must be invalidated or updated. An `on-context-set` handler can be defined which performs whatever operations are needed to reconcile the evaluated flags with the new context. The OpenFeature SDK calls this handler when the global context is modified.
```

The interface shown below has `onContextChange` as optional. I believe this is correct because not every provide will need it and it maintains backwards compatibility.